src/map/instance.cpp

```diff
@@ -809,352 +827,415 @@ enum e_instance_enter instance_enter(struct map_session_data *sd, unsigned short
-	bool isNew = false;
-
-	if (!id || id >  USHRT_MAX || *ptr) {
-		ShowError("instance_readdb_sub: Cannot add instance with ID '%d'. Valid IDs are 1 ~ %d, skipping...\n", id, USHRT_MAX);
+	uint16 instance_id = 0;
+	std::string name;
+	bool existing = false;
+
+	if (!node["ID"]) {
+		yaml_invalid_warning("instance_read_db_sub: Missing ID field in '" CL_WHITE "%s" CL_RESET "', skipping.\n", node, source);
+		return false;
+	}
+	try {
+		instance_id = node["ID"].as<uint16>();
+	} catch (...) {
+		yaml_invalid_warning("instance_read_db_sub: Instance definition with invalid ID field in '" CL_WHITE "%s" CL_RESET "', skipping.\n", node, source);
 		return false;
 	}
 
-	if (mapindex_name2id(str[4]) == 0) {
-		ShowError("instance_readdb_sub: Invalid map '%s' as entrance map, skipping...\n", str[4]);
+	if (instance_id < 1 || instance_id > UINT16_MAX) {
+		ShowWarning("instance_read_db_sub: Invalid instance ID %d in \"%s\", entry #%d (min: 1, max: %d), skipping.\n", instance_id, source.c_str(), n, UINT16_MAX);
 		return false;
 	}
 
-	if (!(db = (struct instance_db *)uidb_get(InstanceDB, id))) {
-		CREATE(db, struct instance_db, 1);
-		db->id = id;
-		db->name = StringBuf_Malloc();
-		db->enter.mapname = StringBuf_Malloc();
-		isNew = true;
-	} else {
-		StringBuf_Clear(db->name);
-		StringBuf_Clear(db->enter.mapname);
-		if (db->maplist_count) {
-			for (i = 0; i < db->maplist_count; i++)
-				StringBuf_Free(db->maplist[i]);
-			aFree(db->maplist);
-			db->maplist = NULL;
+	if (instance_exists(instance_id)) {
+		if (source.find("import") != std::string::npos) // Import file read-in, free previous value and store new value
+			existing = true;
+		else { // Normal file read-in
+			ShowWarning("instance_read_db_sub: Duplicate instance %d, skipping.\n", instance_id);
+			return false;
 		}
-		db->maplist_count = 0;
 	}
 
-	StringBuf_AppendStr(db->name, str[1]);
-
-	db->limit = strtol(str[2], &ptr, 10);
-	if (*ptr) {
-		ShowError("instance_readdb_sub: TimeLimit must be an integer value for instance '%d', skipping...\n", id);
-		instance_db_free_sub(db);
+	if (!node["Name"]) {
+		ShowWarning("instance_read_db_sub: Missing instance name for instance %d in \"%s\", skipping.\n", instance_id, source.c_str());
 		return false;
 	}
+	try {
+		name = node["Name"].as<std::string>();
 
-	db->timeout = strtol(str[3], &ptr, 10);
-	if (*ptr) {
-		ShowError("instance_readdb_sub: IdleTimeOut must be an integer value for instance '%d', skipping...\n", id);
-		instance_db_free_sub(db);
+		if (instance_search_db_name(name.c_str()) != nullptr) {
+			ShowWarning("instance_read_db_sub: Instance name %s already exists for instance %d in \"%s\", skipping.\n", name.c_str(), instance_id, source.c_str());
+			return false;
+		}
+	} catch (...) {
+		yaml_invalid_warning("instance_read_db_sub: Instance definition with invalid name field in '" CL_WHITE "%s" CL_RESET "', skipping.\n", node, source);
 		return false;
 	}
 
-	StringBuf_AppendStr(db->enter.mapname, str[4]);
+	if (!existing)
```

Comment:
I don't get 976 vs 978.
If not existing we adding it then we research it again ?
Why not searching it then adding it if not exist instead ?

auto entry = instance_search_db(instance_id);
if(entry == nullptr) {
 entry = instance_db[instance_id]; //added and shared_ptr copied (ref increased)
}